orderer/common/multichannel/registrar.go

```diff
@@ -1014,31 +1058,44 @@ func (r *Registrar) removeSystemChannel() error {
 			return errors.WithMessagef(err, "failed to determine channel membership for channel: %s", channel)
 		}
 		if !isChannelMember {
-			logger.Debugf("Not a member of channel %s, removing it", channel)
-			err := r.removeMember(channel, cs)
+			logger.Debugf("not a member of channel %s, removing it", channel)
+			err := r.ledgerFactory.Remove(channel)
 			if err != nil {
-				return errors.WithMessagef(err, "failed to remove channel: %s", channel)
+				logger.Errorf("failed removing ledger for channel %s, error: %v", channel, err)
+				failedRemovals = append(failedRemovals, channel)
+				continue
 			}
+
+			delete(r.chains, channel)
 		}
 	}
 
-	// remove system channel resources
-	err := r.ledgerFactory.Remove(systemChannelID)
-	if err != nil {
-		return errors.WithMessagef(err, "error removing ledger for system channel %s", r.systemChannelID)
+	if len(failedRemovals) > 0 {
+		return fmt.Errorf("failed removing ledger for channel(s): %s", strings.Join(failedRemovals, ", "))
 	}
 
-	// remove system channel references
-	r.systemChannel = nil
-	r.systemChannelID = ""
-
 	// reintialize the registrar to recreate every channel
 	r.init(r.consenters)
 
 	// restart every channel
 	r.startChannels()
 
-	logger.Infof("Removed system channel: %s", systemChannelID)
-
 	return nil
 }
+
+func (r *Registrar) removeLedgerAsync(channelID string) {
+	r.pendingRemoval[channelID] = true
+
+	go func() {
+		if err := r.ledgerFactory.Remove(channelID); err != nil {
```

Comment:
can't we make this simpler? 

```
err := r.ledgerFactory.Remove(channelID)
r.lock.Lock()
defer r.lock.Unlock()
if err != nil { 
.... 
}
```

Reply:
very much so yes - I'll handle in a follow-up.